File: removeZeroSum.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def removeZeroSumSublists(self, head: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        curr = dummy
        prefixMap = {}
        prefixSum = 0

        while curr:
            prefixSum += curr.val
            if prefixSum in prefixMap:
                prev = prefixMap[prefixSum]
                curr = prev.next

                p = prefixSum + curr.val
                while p != prefixSum:
                    del prefixMap[p]
                    curr = curr.next
                    p += curr.val
                
                prev.next = curr.next
            else:
                prefixMap[prefixSum] = curr
            curr = curr.next
        
        return dummy.next

File: removeZeroSum.java
import java.util.HashMap;
import java.util.Map;

public class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}

public class removeZeroSum {
    public ListNode removeZeroSumSublists(ListNode head) {
        ListNode dummy = new ListNode(0, head);
        ListNode curr = dummy;
        Map<Integer, ListNode> prefixMap = new HashMap<>();
        int prefixSum = 0;

        while (curr != null) {
            prefixSum += curr.val;
            if (prefixMap.containsKey(prefixSum)) {
                ListNode prev = prefixMap.get(prefixSum);
                curr = prev.next;

                int p = prefixSum + curr.val;
                while (p != prefixSum) {
                    prefixMap.remove(p);
                    curr = curr.next;
                    p += curr.val;
                }

                prev.next = curr.next;
            } else {
                prefixMap.put(prefixSum, curr);
            }
            curr = curr.next;
        }

        return dummy.next;
    }

    public static void main(String[] args) {
        
    }
}
